### ISSUE TITLE
Not syncing when using only static nodes

### ISSUE THREAD

**Issue by darkFunction** [ConsenSys/quorum-kubernetes#173]
I'm a bit confused about the conflicting information in the documentation about bootnodes and static nodes in a private network. I understand from [here](https://besu.hyperledger.org/en/stable/private-networks/how-to/configure/bootnodes/) that I can use only static nodes, also from the [README](https://github.com/ConsenSys/quorum-kubernetes/tree/master/helm#readme) here that bootnodes are optional.

However I also see:

> A network must have at least one operating bootnode
> In private networks for development or testing purposes, specify at least one bootnode.

I've been trying to get the basic helm deployment working in minikube without using bootnodes (just validators 1-4) and have been unable to get them syncing. I've had success outside kubernetes launching nodes locally and specifying bootnodes, but I'm running on constrained resources for my k8s cluster and would rather just have 4 validators than spinning up extra nodes to act as bootnodes when I know in advance all the nodes which are participating in my test network.

I've double checked the node init logs and can see that the 4 enode public keys match the ones in the `besu-peers` ConfigMap and `static-nodes.json`.

I am using both node and address contract permissioning. Does this mean I need to include the static enodes in the genesis file?

**Comment by darkFunction** [comment]
I've noticed NAT warnings and after forcing the nat-method to KUBERNETES I now see startup failure:

```
2022-12-06 19:09:01.840+00:00 | main | INFO  | Runner | Starting Ethereum main loop ... 
2022-12-06 19:09:01.840+00:00 | main | INFO  | KubernetesNatManager | Starting kubernetes NAT manager.
Startup failed
To display full help:
besu [COMMAND] --help
2022-12-06 19:09:02.798+00:00 | main | WARN  | NatService | Nat manager failed to configure itself automatically due to the following reason : org.hyperledger.besu.nat.core.exception.NatInitializationException: . 
2022-12-06 19:09:02.798+00:00 | main | ERROR | Runner | unable to start main loop
java.lang.IllegalStateException: 
 at org.hyperledger.besu.nat.NatService.start(NatService.java:107)
 at org.hyperledger.besu.Runner.startEthereumMainLoop(Runner.java:138)
 at org.hyperledger.besu.cli.BesuCommand.run(BesuCommand.java:1453)
```

**Comment by joshuafernandes** [comment]
Hi @darkFunction 

Could you please give us a list of steps to reproduce this error you are seeing? Which client are you using and what your environment is like. If you've made changes to the values.yml please list those out too.

Cheers

**Comment by joshuafernandes** [comment]
> I'm a bit confused about the conflicting information in the documentation about bootnodes and static nodes in a private network. I understand from [here](https://besu.hyperledger.org/en/stable/private-networks/how-to/configure/bootnodes/) that I can use only static nodes, also from the [README](https://github.com/ConsenSys/quorum-kubernetes/tree/master/helm#readme) here that bootnodes are optional.
> 
> However I also see:
> 
> > A network must have at least one operating bootnode
> > In private networks for development or testing purposes, specify at least one bootnode.
> 
> I've been trying to get the basic helm deployment working in minikube without using bootnodes (just validators 1-4) and have been unable to get them syncing. I've had success outside kubernetes launching nodes locally and specifying bootnodes, but I'm running on constrained resources for my k8s cluster and would rather just have 4 validators than spinning up extra nodes to act as bootnodes when I know in advance all the nodes which are participating in my test network.
> 
> I've double checked the node init logs and can see that the 4 enode public keys match the ones in the `besu-peers` ConfigMap and `static-nodes.json`.
> 
> I am using both node and address contract permissioning. Does this mean I need to include the static enodes in the genesis file?

For networks in general you used need bootnodes - the same holds true for public networks like mainnet and private networks you create. A little while ago Besu added static-nodes.json to its config (in 2021 I believe) so now you can use either on private networks. The order of things when a node fires up is:
- check for bootnodes and get a list of peers and attempt to connect to each and peer
- if this isn't available it will look for static-nodes.json and attempt to connect to those 
- if neither is found, it'll error
Basically, when a node comes up it needs to connect to other nodes and peer - its happy if it gets a list from either bootnodes or from static nodes. 

That being said it is still a good idea to have bootnodes especially when you have a wider private network or other participants. You are generally not aware of every single peer on the network, but they all connect to the bootnodes so the bootnodes maintain a good list of peers. When you are running the network on minikube or aks and it's just you using a handful of nodes it isn't really needed. With the charts here they automatically update the peers list which is effectively the same thing

Hope that helps

**Comment by darkFunction** [comment]
Thank you for your detailed response.

I don't wish to have bootnodes as I'm operating a confined and experimental private network, where I am aware of all nodes.

> Your network may not be working atm because of the contract permissioning pieces.

This was the problem- when I disable node permissioning, then the nodes find each other. But I'm not sure why- my understanding is that the list of static nodes overrules the node permissioning- otherwise there's no way to get a blockchain up and running in order to deploy the node permission contract.

**Comment by darkFunction** [comment]
I've also tried updating the NodeIngress contract to always return `true` for `connectionAllowed(...)`, and when using *TRACE* log level I see this:

```
nioEventLoopGroup-3-7 | TRACE | NodePermissioningController | Node permissioning - Sync Status: Rejected enode://...
```

**Comment by joshuafernandes** [comment]
Hi @darkFunction 
> 
> > Your network may not be working atm because of the contract permissioning pieces.
> 
> This was the problem- when I disable node permissioning, then the nodes find each other. But I'm not sure why- my understanding is that the list of static nodes overrules the node permissioning- otherwise there's no way to get a blockchain up and running in order to deploy the node permission contract.

Permissions is the highest level so to speak, static nodes just helps the nodes peer ie at boot the node can connect to other peers on the network. This is similar to bootnodes ie contact 1-2 bootnodes -> discover 10-25 peers -> UDP packet to discover each to see if they still exist -> TCP connection to each -> peer and gossip. Static nodes skips the first few steps and directly gives it the list of peers so it sends the UDP packet to each -> TCP connection to each -> peer and gossip. When you use permissions you still need to have the list of static peers in the permissions list else the node attempts to connect to each peer -> checks permissions -> fails if it isn't in the list

Rejecting `enode...` means its not allowed to connect to the peer - sometimes this is because the IP is reported back differently (quite often the case on k8s due to the complex networking layers). The note re bootnodes just means that until a node is synced it will not create blocks (if its a validator). Here's a video tutorial (its a couple year's old but the concepts are shown there https://www.youtube.com/watch?v=MhOJKOoEZQQ)

Cheers
Josh

**Comment by darkFunction** [comment]
Thanks Josh- I'll dig into the IP a bit more, I see IP of 0.0.0.0 on enode addresses. My understanding of kubernetes is still limited! This seems like a good avenue to investigate

**Comment by darkFunction** [comment]
Here's the relevant parts of my logs with a breakdown of what's happening:

There are 4 nodes on the network. They are accurately represented in the `besu-peers` static nodes ConfigMap. Nodes add their own address into this list during the hook `node-hooks-pre-install` but the old entries are left in. I don't think this matters and have redacted the old entries below. I've also verified that these keys map correctly to the addresses in the `besu-node-validator-x-address` ConfigMaps.

```
"static-nodes.json": "[
	  <unused>,
	  <unused>,
	  \"enode://77ebe...6edf0@besu-node-validator-3-0.besu-node-validator-3.quorum.svc.cluster.local:30303?discport=0\",
	  \"enode://78ff4...c1897@besu-node-validator-2-0.besu-node-validator-2.quorum.svc.cluster.local:30303?discport=0\",
	  \"enode://7caea...7fce6@besu-node-validator-4-0.besu-node-validator-4.quorum.svc.cluster.local:30303?discport=0\",
	  <unused>,
	  \"enode://ae8904...77892@besu-node-validator-1-0.besu-node-validator-1.quorum.svc.cluster.local:30303?discport=0\",
	  <unused>
]
```

Let's have a look at the logs for Node 4:

```
 Connecting to 8 static nodes.
 Static Nodes = [enode://14684...0cd60@besu-node-validator-3-0.besu-node-validator-3.quorum.svc.cluster.local:30303?discport=0, enode://                                                            78ff48...c1897@besu-node-validator-2-0.besu-node-validator-2.quorum.svc.cluster.local:30303?discport=0, enode://7caea...7fce6@besu-    node-validator-4-0.besu-node-validator-4.quorum.svc.cluster.local:30303?discport=0, enode://77ebe9...6edf0@besu-node-validator-3-0.besu-node-validator-3.quorum.svc.cluster.local:30303?discport=0, enode://ae8904...77892@besu-node-validator-1-0.besu-node-validator-1.quorum.svc.cluster.local:30303?discport=0, enode://74b41...4cfe3@besu-    node-validator-4-0.besu-node-validator-4.quorum.svc.cluster.local:30303?discport=0, enode://eaf366...1953b@besu-node-validator-1-0.besu-node-validator-1.quorum.svc.cluster.local:30303?discport=0, enode://886bd...9477a@besu-node-validator-2-0.besu-node-validator-2.quorum.svc.cluster.local:30303?discport=0]
```

Looks good so far.

```
 NodePermissioningControllerFactory | Validating onchain node permissioning smart contract configuration
 NodePermissioningController | Node permissioning: Checking enode://6f8a80...d92a0@10.3.58.6:30303 -> enode://                                                                                              6f8a80...d92a0@10.3.58.6:30303
 MainnetTransactionProcessor | Starting execution of MessageCall
...
 MessageCallProcessor | Successful message call of 0x0000000000000000000000000000000000000000 to 0x0000000000000000000000000000000000009999 (Gas remaining: 4667396)
 MainnetTransactionProcessor | Gas used by transaction: 32604, by message call/contract creation: 900
 NodeSmartContractV2PermissioningController | Permitted? true for IP enode://6f8a80...d92a0@10.3.58.6:30303
```

OK, don't really understand where this enode comes from but presuming it's a hardcoded test enode for verifying contract functionality. Seems to be happy that it's working. Note it returns `true` here because I wrote the NodeIngress to always  return true for testing. Startup is good and we have the Node-4 enode:

```
 | main | INFO  | DefaultP2PNetwork | Enode URL enode://7caeaf1c34d591b1f107f17fa6275295f77b4a472c6d7a49afb75de513aeb4f06d06dc51adf58d2c732c736195f99678841d5516f99146e289076d791917fce6@0.0.0.0:30303?discport=0
```

Now we try to communicate with Node 3 (77ebe9):

```
  AbstractHandshakeHandler | Sending framed hello
  AbstractPeerConnection | Writing DisconnectMessage{data=0xc180} to PeerInfo{version=5, clientId='besu/v22.7.2/linux-x86_64/openjdk-java-11', capabilities=[eth/62, eth/63, eth/64, eth/65, eth/66, istanbul/100, snap/1], port=30303, nodeId=0x78ff48...cc1897} via protocol null
  AbstractHandshakeHandler | Successfully wrote hello message
  DeFramer | Received HELLO message: PeerInfo{version=5, clientId='besu/v22.7.2/linux-x86_64/openjdk-java-11', capabilities=[eth/62, eth/63, eth/64, eth/65, eth/66, istanbul/100, snap/1], port=30303, nodeId=0x77ebe9...06edf0}
  DeFramer | Enable compression for p2pVersion: 5
  NodePermissioningController | Node permissioning: Checking enode://77ebe9...6edf0@172.17.0.15:30303?discport=0 -> enode://7caea...17fce6@0.0.0.0:30303?discport=0
  NodePermissioningController | Node permissioning - Sync Status: Rejected enode://77ebe9...906edf0@172.17.0.15:30303?discport=0 -> enode://7caea...7fce6@0.0.0.0:30303?discport=0
  RlpxAgent | Node is not permitted to connect. Disconnect incoming connection: NettyPeerConnection{nodeId=0x77ebe9...06edf0, clientId=besu/v22.7.2/linux-x86_64/openjdk-java-11, caps=eth/66, istanbul/100, snap/1}
```

I can see that the address is different for each node, for the local node 4 it's 0.0.0.0 and the remote node 3 it's 172.17.0.15, as opposed to what is in the static nodes list (`besu-node-validator-3-0.besu-node-validator-3.quorum.svc.cluster.local`).

Is this the issue? Why would it be working when I *disable* node permissioning if this were the case?

**Comment by joshuafernandes** [comment]
> Thanks @joshuafernandes. I watched the video and it was very useful. I noticed that you spun your network up first before activating node permissioning on any of the nodes, you said it was a chicken-and-egg problem to try to bootstrap a new network with node permissions enabled. Wondering if I should be doing the same.
> 
> Here's the relevant parts of my logs with a breakdown of what's happening:
> 
> There are 4 nodes on the network. They are accurately represented in the `besu-peers` static nodes ConfigMap. Nodes add their own address into this list during the hook `node-hooks-pre-install` but the old entries are left in. I don't think this matters and have redacted the old entries below. I've also verified that these keys map correctly to the addresses in the `besu-node-validator-x-address` ConfigMaps.
> 
> ```
> "static-nodes.json": "[
> 	  <unused>,
> 	  <unused>,
> 	  \"enode://77ebe...6edf0@besu-node-validator-3-0.besu-node-validator-3.quorum.svc.cluster.local:30303?discport=0\",
> 	  \"enode://78ff4...c1897@besu-node-validator-2-0.besu-node-validator-2.quorum.svc.cluster.local:30303?discport=0\",
> 	  \"enode://7caea...7fce6@besu-node-validator-4-0.besu-node-validator-4.quorum.svc.cluster.local:30303?discport=0\",
> 	  <unused>,
> 	  \"enode://ae8904...77892@besu-node-validator-1-0.besu-node-validator-1.quorum.svc.cluster.local:30303?discport=0\",
> 	  <unused>
> ]
> ```
> 
> Let's have a look at the logs for Node 4:
> 
> ```
>  Connecting to 8 static nodes.
>  Static Nodes = [enode://14684...0cd60@besu-node-validator-3-0.besu-node-validator-3.quorum.svc.cluster.local:30303?discport=0, enode://                                                            78ff48...c1897@besu-node-validator-2-0.besu-node-validator-2.quorum.svc.cluster.local:30303?discport=0, enode://7caea...7fce6@besu-    node-validator-4-0.besu-node-validator-4.quorum.svc.cluster.local:30303?discport=0, enode://77ebe9...6edf0@besu-node-validator-3-0.besu-node-validator-3.quorum.svc.cluster.local:30303?discport=0, enode://ae8904...77892@besu-node-validator-1-0.besu-node-validator-1.quorum.svc.cluster.local:30303?discport=0, enode://74b41...4cfe3@besu-    node-validator-4-0.besu-node-validator-4.quorum.svc.cluster.local:30303?discport=0, enode://eaf366...1953b@besu-node-validator-1-0.besu-node-validator-1.quorum.svc.cluster.local:30303?discport=0, enode://886bd...9477a@besu-node-validator-2-0.besu-node-validator-2.quorum.svc.cluster.local:30303?discport=0]
> ```
> 
> Looks good so far.
> 
> ```
>  NodePermissioningControllerFactory | Validating onchain node permissioning smart contract configuration
>  NodePermissioningController | Node permissioning: Checking enode://6f8a80...d92a0@10.3.58.6:30303 -> enode://6f8a80...d92a0@10.3.58.6:30303
>  MainnetTransactionProcessor | Starting execution of MessageCall
> ...
>  MessageCallProcessor | Successful message call of 0x0000000000000000000000000000000000000000 to 0x0000000000000000000000000000000000009999 (Gas remaining: 4667396)
>  MainnetTransactionProcessor | Gas used by transaction: 32604, by message call/contract creation: 900
>  NodeSmartContractV2PermissioningController | Permitted? true for IP enode://6f8a80...d92a0@10.3.58.6:30303
> ```
> 
> OK, don't really understand where this enode comes from but presuming it's a hardcoded test enode for verifying contract functionality. Seems to be happy that it's working. Note it returns `true` here because I wrote the NodeIngress to always return true for testing. Startup is good and we have the Node-4 enode:
> 

Something doesn't look right here, you may need to jump into the container and see what key is being used here - the IP address can change but the pub key part of the enode should not, I think what's happening is that it either can't find the correct key (path may be incorrect) and so creates a new key for itself; hence the new enode. Any new enode like so will not work with the others as below coz its disallowed. 

I think the better approach for this is to use the templates in the playground section which have hard coded keys and you can see what happpens when you change values there when you add permissions. Once you have the idea switch to the helm charts?

> ```
>  | main | INFO  | DefaultP2PNetwork | Enode URL enode://7caeaf1c34d591b1f107f17fa6275295f77b4a472c6d7a49afb75de513aeb4f06d06dc51adf58d2c732c736195f99678841d5516f99146e289076d791917fce6@0.0.0.0:30303?discport=0
> ```
> 
> Now we try to communicate with Node 3 (77ebe9):
> 
> ```
>   AbstractHandshakeHandler | Sending framed hello
>   AbstractPeerConnection | Writing DisconnectMessage{data=0xc180} to PeerInfo{version=5, clientId='besu/v22.7.2/linux-x86_64/openjdk-java-11', capabilities=[eth/62, eth/63, eth/64, eth/65, eth/66, istanbul/100, snap/1], port=30303, nodeId=0x78ff48...cc1897} via protocol null
>   AbstractHandshakeHandler | Successfully wrote hello message
>   DeFramer | Received HELLO message: PeerInfo{version=5, clientId='besu/v22.7.2/linux-x86_64/openjdk-java-11', capabilities=[eth/62, eth/63, eth/64, eth/65, eth/66, istanbul/100, snap/1], port=30303, nodeId=0x77ebe9...06edf0}
>   DeFramer | Enable compression for p2pVersion: 5
>   NodePermissioningController | Node permissioning: Checking enode://77ebe9...6edf0@172.17.0.15:30303?discport=0 -> enode://7caea...17fce6@0.0.0.0:30303?discport=0
>   NodePermissioningController | Node permissioning - Sync Status: Rejected enode://77ebe9...906edf0@172.17.0.15:30303?discport=0 -> enode://7caea...7fce6@0.0.0.0:30303?discport=0
>   RlpxAgent | Node is not permitted to connect. Disconnect incoming connection: NettyPeerConnection{nodeId=0x77ebe9...06edf0, clientId=besu/v22.7.2/linux-x86_64/openjdk-java-11, caps=eth/66, istanbul/100, snap/1}
> ```
> 
> I can see that the address is different for each node, for the local node 4 it's 0.0.0.0 and the remote node 3 it's 172.17.0.15, as opposed to what is in the static nodes list (`besu-node-validator-3-0.besu-node-validator-3.quorum.svc.cluster.local`).
> 
> Is this the issue? Why would it be working when I _disable_ node permissioning if this were the case? And I've overriden the `allowsConnection` contract function to always return true, so it doesn't matter what address gets passed in here.

**Comment by darkFunction** [comment]
> Something doesn't look right here, you may need to jump into the container and see what key is being used here - the IP address can change but the pub key part of the enode should not, I think what's happening is that it either can't find the correct key (path may be incorrect) and so creates a new key for itself; hence the new enode. Any new enode like so will not work with the others as below coz its disallowed.

> I think the better approach for this is to use the templates in the playground section which have hard coded keys and you can see what happpens when you change values there when you add permissions. Once you have the idea switch to the helm charts?

Thanks and I appreciated the detailed follow up. I took your advice and spent a lot more time digging into how nodes generate and store public keys, and made some good progress.

My findings:

 - `besu-genesis-init` Job is correctly setting up my `besu-peers` and all the Secrets containing the enode information, pub key etc. 
 - what was confusing the matter was the pre-install hook on besu node installs- it was generating a new enode, overwriting the Secrets and adding itself to `besu-peers`. So each node was starting with a new enode which wasn't added to the `static-nodes.json` of previously installed nodes.
 - I simplified by just getting rid of the pre-install hook, I don't need it. So now I have a static set of enodes in`besu-peers` and in Secrets and everything is easy to follow
 - verified that I can tear down all nodes and install a new set with same set of enodes and everything works again (nodePermissions disabled)

Now I change `nodesContract -> enabled` to `true` and spin nodes up, they fail. I have a trace log of node 1 (attached), and everything looks OK to my untrained eye, except I see `TCP_SUBSYSTEM_ERROR` and some disconnects, which look concerning.

All the node info in the logs is correct now, and all the nodes are attempting to communicate using the correct public keys at least. 

Would very much appreciate if you could cast an eye over the log- will drop this in Discord too!

[logs-from-validator-1-besu-in-besu-node-validator-1-0.log](https://github.com/ConsenSys/quorum-kubernetes/files/10406175/logs-from-validator-1-besu-in-besu-node-validator-1-0.log)

**Comment by darkFunction** [comment]
I've found the issue!

I was unable to bootstrap a network with node permissioning enabled, but could bootstrap with it disabled and subsequently enable- everything worked fine after this.

eg,
`for i in {1..4}; do helm upgrade validator-$i ./charts/besu-node --namespace quorum --values ./values/validator.yml --set node.besu.permissions.nodesContract.enabled=true; done`

**Comment by darkFunction** [comment]
Actually, it was a red herring, my mistake. I was upgrading node config but not restarting.

When I restart the nodes with permissioning enabled, I see this:

"Unable to create ECDH Key agreement due to Crypto engine failure"

Have verified that priv/public keys and addresses are correct on each node and indeed when I relaunch with the only change being  `--set node.besu.permissions.nodesContract.enabled=true` then they again fail to sync with the above error.

When I restart the nodes with `nodeContract.enabled=false` then they sync up again no problem.

**Comment by joshuafernandes** [comment]
Hi @darkFunction thanks for this. I need to have a look at this when I can get a chance depending on priorities, given that this is working fine sans permissions, this looks more like an issue with the contracts or the setup /docs we have is perhaps out of date - could you verify this works the same with https://github.com/ConsenSys/quorum-dev-quickstart (easier to replicate there rather than here on a cluster for the other developers) and raise a ticket on https://github.com/hyperledger/besu/issues please?
